File: foundations/network-scanning/windows-airgapped-ip-discovery/discover_ips.go
package main

import (
	"fmt"
	"net"
	"os"
	"runtime"
	"sync"
	"time"
)

func pingHost(ip string) bool {
	timeout := 350 * time.Millisecond
	conn, err := net.DialTimeout("tcp", net.JoinHostPort(ip, "135"), timeout)
	if err == nil {
		_ = conn.Close()
		return true
	}
	return false
}

func main() {
	cidr := "192.168.10.0/24"
	if len(os.Args) > 1 {
		cidr = os.Args[1]
	}

	ip, ipNet, err := net.ParseCIDR(cidr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "invalid CIDR: %v\n", err)
		os.Exit(1)
	}

	base := ip.To4()
	if base == nil || ipNet.Mask.String() != "ffffff00" {
		fmt.Fprintln(os.Stderr, "only IPv4 /24 CIDR is supported in this starter script")
		os.Exit(1)
	}

	prefix := fmt.Sprintf("%d.%d.%d", base[0], base[1], base[2])
	workers := runtime.NumCPU()
	jobs := make(chan string)
	results := make(chan string)
	var wg sync.WaitGroup

	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for host := range jobs {
				if pingHost(host) {
					results <- host
				}
			}
		}()
	}

	go func() {
		for i := 1; i <= 254; i++ {
			jobs <- fmt.Sprintf("%s.%d", prefix, i)
		}
		close(jobs)
		wg.Wait()
		close(results)
	}()

	for host := range results {
		fmt.Println(host)
	}
}

File: foundations/network-scanning/windows-airgapped-ip-discovery/discover_ips.rs
use std::env;
use std::process::Command;

fn ping_host(ip: &str) -> bool {
    let output = if cfg!(target_os = "windows") {
        Command::new("ping").args(["-n", "1", "-w", "300", ip]).output()
    } else {
        Command::new("ping").args(["-c", "1", "-W", "1", ip]).output()
    };

    output.map(|o| o.status.success()).unwrap_or(false)
}

fn main() {
    let cidr = env::args().nth(1).unwrap_or_else(|| "192.168.10.0/24".to_string());
    let parts: Vec<&str> = cidr.split('/').collect();

    if parts.len() != 2 || parts[1] != "24" {
        eprintln!("Only IPv4 /24 CIDR is supported in this starter script.");
        std::process::exit(1);
    }

    let octets: Vec<&str> = parts[0].split('.').collect();
    if octets.len() != 4 {
        eprintln!("Invalid IPv4 network base.");
        std::process::exit(1);
    }

    let prefix = format!("{}.{}.{}", octets[0], octets[1], octets[2]);
    for host in 1..=254 {
        let ip = format!("{}.{}", prefix, host);
        if ping_host(&ip) {
            println!("{}", ip);
        }
    }
}
